### DISCUSSION TITLE
Idea: "value hooks" (for lack of better wording)

### DISCUSSION THREAD

**Discussion by joakimbeng** [css-hooks/css-hooks#44] [Ideas]
The current "hooks" only hook into the property names in a style object. How about making it possible to hook into (and replace) property values as well? I'm thinking as a way for having a consistent design system.

An example is better than a few words:

```tsx
<div style={css({
  color: "green-500",
  hover: {
    color: "green-400",
  },
  padding: "s-4",
})}/>
```

So when css hooks is configured in a project you can provide "value hooks" or substitutions. On top of my head it's mainly for spacings and colors so one could probably do something smart with TypeScript as well to get autocomplete for such values inside the properties that they apply to, e.g colors should only be used for text, background, border, and outline color (if I haven't forgotten about some other properties).

This way it would be as easy as with Tailwind to just using a color or similar anywhere, without having to import some design token all the time.

What do you think?

**Comment by nsaunders** [comment]
Hi @joakimbeng! This is a great idea, but a bit out-of-scope for this library.

But since you own the `css` function CSS Hooks gives you, you could always create a wrapper that would replace the design token values for you. Take this (untested pseudo-code) for example:

```typescript
// css-hooks.ts

const [hooks, applyHooks] = createHooks({
  // hook configuration
});

function replaceTokens<O extends Record<string, unknown>>(properties: O): O {
  // todo: traverse style object, replacing design token values
  // then return it
}

export const css: typeof applyHooksToStyleObject = properties => applyHooks(replaceTokens(properties));

export { hooks }
```

Personally, I have a much lower-tech approach that seems to work well for me so far, along the lines of _CSS variable references encoded in TypeScript variables_. For example:

```typescript
// design-tokens.ts

export const green400 = "var(--green-400, #0b0)";
export const green500 = "var(--green-500, #090)";
export const spacing4 = "var(--spacing-4, 1rem)";
// etc.
```

Usage:

```tsx
import * as V from "./design-tokens";

<button
  style={css({
    padding: V.spacing4,
    color: V.green500,
    hover: {
      color: V.green400
    }
  })} />
```

Simplicity aside, this approach has a few advantages:
1. type safety / constraints (which you could potentially take further, e.g. with namespacing)
2. great autocompletion (thanks to the previous point really)
3. ability to change the theme at runtime via CSS variables
4. pretty much the DX you're looking for, I think? (Please correct me if I'm mistaken.)

Whether built on one of these two approaches or a different pattern entirely, I do think there's plenty of room for more of an opinionated "CSS framework" built on inline styles and CSS Hooks. Actually, I got started on something for my own use ([Varsace](https://github.com/nsaunders/varsace) - just color palette for now); but I'm too short on time (and possibly too unopinionated) to carry this forward right at the moment. But, if that does strike any interest for you, please let me know and I'd be happy to collaborate.

Either way, really glad to have you aboard Joakim, and thanks for sharing your thoughts. Please keep your feedback coming!

**Comment by joakimbeng** [comment]
Thank's for the good answer! I'll tinker with it in userland ☺️ 

Regarding the DX I'm looking for it's something similar to Tailwind, where I don't have to import anything for it to work. But the benefit of using css-hooks and my "value hooks" with good TypeScript types over Tailwind is that it's easier to configure than Tailwind and I don't need any extensions in my IDE to get intellisense for the different colors.

The DX I'm thinking about is something similar to:

* I'm typing `style={css` <kbd>&#8595;</kbd>+<kbd>enter</kbd> (or whatever the key combo is to get VS Code to auto-import the `css` function)
* then when I continue to type `({ color: '` I'll get autocomplete for the colors in my project

Will play around with it to see how it would work and feel 😊

**Comment by woodcox** [comment]
Have you come across an older css project called [bonsai.css](https://github.com/bonsaicss/bonsai.css/blob/master/dist/bonsai-utilities.css)? The utility class aspect of the library was very interesting use of custom properties in the style tag:

~~~css
[style*="--bg:"] {
    background: var(--bg);
}
~~~

This allowed for the following Style based utilities:

~~~html
<div style="--bg:var(--dark)"><div>
<div style="--bg:#2A2A2A"><div>
<div style="--bg:rgb(42,42,42)"><div>
<div style="--bg:rgba(0,0,0,.8)"><div>
~~~

It even allowed for responsive design when combined with media queries. Sadly the website has been removed but the docs repo still exists on [GitHub](https://github.com/ciar4n/bonsai-docs). 

This utility class system may be out of scope for you, but it seems like a really good fit.
